src/plugins/connectionPlugin.ts

```diff
@@ -0,0 +1,983 @@
+                  name: edgeName,
+                  definition(t2) {
+                    t2.string("cursor", {
+                      nullable: false,
+                      description:
+                        "https://facebook.github.io/relay/graphql/connections.htm#sec-Cursor",
+                    });
+
+                    t2.field("node", {
+                      type: targetType,
+                      description:
+                        "https://facebook.github.io/relay/graphql/connections.htm#sec-Node",
+                    });
+
+                    if (pluginExtendEdge) {
+                      eachObj(pluginExtendEdge, (val, key) => {
+                        t2.field(key, val);
+                      });
+                    }
+
+                    if (fieldConfig.extendEdge instanceof Function) {
+                      fieldConfig.extendEdge(t2);
+                    }
+                  },
+                })
+              );
+            }
+
+            // Add the "PageInfo" type to the schema if it doesn't exist already
+            if (!b.hasType("PageInfo")) {
+              b.addType(
+                objectType({
+                  name: "PageInfo",
+                  description:
+                    "PageInfo cursor, as defined in https://facebook.github.io/relay/graphql/connections.htm#sec-undefined.PageInfo",
+                  definition(t2) {
+                    t2.boolean("hasNextPage", {
+                      nullable: false,
+                      description: `Used to indicate whether more edges exist following the set defined by the clients arguments.`,
+                    });
+                    t2.boolean("hasPreviousPage", {
+                      nullable: false,
+                      description: `Used to indicate whether more edges exist prior to the set defined by the clients arguments.`,
+                    });
+                    t2.string("startCursor", {
+                      nullable: true,
+                      description: `The cursor corresponding to the first nodes in edges. Null if the connection is empty.`,
+                    });
+                    t2.string("endCursor", {
+                      nullable: true,
+                      description: `The cursor corresponding to the last nodes in edges. Null if the connection is empty.`,
+                    });
+                  },
+                })
+              );
+            }
+            const {
+              disableBackwardPagination,
+              disableForwardPagination,
+              validateArgs = defaultValidateArgs,
+            } = {
+              ...pluginConfig,
+              ...fieldConfig,
+            };
+
+            let specArgs = {};
+            if (disableForwardPagination !== true) {
+              specArgs = { ...ForwardPaginateArgs };
+            }
+            if (disableBackwardPagination !== true) {
+              specArgs = { ...specArgs, ...BackwardPaginateArgs };
+            }
```

Comment:
Some Relay-style APIs allow making `first` a non-nullable argument if backward pagination is disabled (since the spec considers the lack of both `first` and `last` to be undefined behavior). Similarly, `last` can be non-nullable if forward pagination is disabled. Maybe this plugin can generate the types like that? Or expose a flag to enable it?

Reply:
That's a good idea - I think a flag (`strictArgs`) that defaults to `true` (enabled) might be the best approach, since some folks may not want this behavior

For instance, in the main project I'm using Nexus on, we currently use a `perPage` / `offset` style forward-pagination. When we add this plugin, we'll want the `first` type to be loose with a custom `validateArgs` fn, so we'll want the flag to disable.

Here's the [commit](https://github.com/prisma-labs/nexus/pull/324/commits/b42c9a6455f571d95ce604ec4fe0c7dab365050e) - any reason you think it'd be better to keep it opt-in?

Reply:
Seems fine this way, thanks for adding it!